lib/ddtrace/configuration/agent_settings_resolver.rb

```diff
@@ -51,27 +66,50 @@ def self.call(settings, logger: Datadog.logger)
         :logger,
         :settings
 
+      attr_writer :default_settings
+
       def initialize(settings, logger: Datadog.logger)
         @settings = settings
         @logger = logger
+        @default_settings = nil
       end
 
       def call
+        # If no agent settings have been provided, we try to connect using a local unix socket.
+        # We only do so if the socket is present when `ddtrace` runs.
+        #
+        # Note we still haven't exposed a way to to configure unix socket connectivity using
+        # environment variables, only using the custom configuration proc.
+        adapter_ = if default_settings && (uds_path_ = uds_path)
+                     Ext::Transport::UnixSocket::ADAPTER
+                   else
+                     # We can't invoke private methods with `self.method` in Ruby < 2.7, thus
+                     # the local variable has to be renamed to avoid conflict.
+                     adapter
+                   end
```

Comment:
I read `AgentSettingsResolver` as the authoritative entity for deciding how to connect to the agent, regardless of product. I was thinking if Security wants to submit data through the agent, it would need this same information and would need to add code around `if uds_path this, else that`, and I thought this decision felt like it was leaking out of `AgentSettingsResolver`.

Reply:
In other words, I see the `AgentSettingsResolver` as the encapsulation of all the complexity around figuring out how to connect to the agent, thus I see choosing the adapter as part of it.
The main reason for that is that whether you'll have to use `uds_path` or `hostname` can be decided by the `AgentSettingsResolver`.

I understand that the Adapter is actually tied to day to implementation (`:net_http` actually means `Net::HTTP` backed today), which I don't think is correct, but returning a value that lets the caller know "you should use TCP+HTTP for this configuration" seems reasonable to me.

Reply:
I could change it from `adapter` to `protocol`, would that make sense to you? Then there's additional logic in each `AgentSettingsResolver` to figure out what adapter to use for the required protocol.

Reply:
Ah, wait, I think my suggestion was misunderstood. I fully agree with all that you mentioned :)

I was mainly referring to code organization inside the `agent_settings_resolver.rb` file: I was saying that inside the `#call` method we could just call to an `#adapter` method, and have this logic there, so that the `#call` method can be 100% focused on just creating the `AgentSettings`, and leaves all of the logic to its helper methods.

I think that otherwise calling it adapter is fine.

Reply:
Refactor done!